src/libraries/System.IO.Compression/src/System/IO/Compression/WinZipAesStream.cs

```diff
@@ -0,0 +1,770 @@
+                finally
+                {
+                    _aesEncryptor?.Dispose();
+                    _aes.Dispose();
+                    _hmac.Dispose();
+                    _encryptionBuffer?.Dispose();
+
+                    // Only dispose the base stream if we don't leave it open
+                    if (!_leaveOpen)
+                    {
+                        _baseStream.Dispose();
+                    }
+                }
+            }
+
+            _disposed = true;
+            base.Dispose(disposing);
+        }
+
+        // Replace DisposeAsync with this version:
+        public override async ValueTask DisposeAsync()
+        {
+            if (_disposed)
+                return;
+
+            try
+            {
+                if (_encrypting && !_authCodeValidated && _headerWritten)
+                {
+                    // make sure header is flushed
+                    await _baseStream.FlushAsync(CancellationToken.None).ConfigureAwait(false);
+
+                    // Encrypt all buffered data first
+                    await FlushEncryptionBufferAsync(CancellationToken.None).ConfigureAwait(false);
+
+                    // Flush any pending data
+                    if (_baseStream.CanWrite)
+                    {
+                        await _baseStream.FlushAsync(CancellationToken.None).ConfigureAwait(false);
+                    }
+
+                    // Write the authentication code (this is still sync, which is fine)
+                    await WriteAuthCodeAsync(CancellationToken.None).ConfigureAwait(false);
+
+                    // Flush again to ensure auth code is written
+                    if (_baseStream.CanWrite)
+                    {
+                        await _baseStream.FlushAsync(CancellationToken.None).ConfigureAwait(false);
+                    }
+                }
+                else if (!_encrypting && !_authCodeValidated && _headerRead)
+                {
+                    // For decryption, validate auth code and CRC if not already done
+                    await ValidateAuthCodeAsync(CancellationToken.None).ConfigureAwait(false);
+                }
+            }
+            finally
+            {
+                _aesEncryptor?.Dispose();
+                _aes.Dispose();
+                _hmac.Dispose();
+                _encryptionBuffer?.Dispose();
+
+                // Only dispose the base stream if we don't leave it open
+                if (!_leaveOpen)
+                {
+                    await _baseStream.DisposeAsync().ConfigureAwait(false);
+                }
+            }
+
+            _disposed = true;
+            GC.SuppressFinalize(this);
+        }
+        public override bool CanRead => !_encrypting && !_disposed;
+        public override bool CanSeek => false;
+        public override bool CanWrite => _encrypting && !_disposed;
+        public override long Length => throw new NotSupportedException();
+        public override long Position
```

Comment:
Position refers to the amount of data read/written by the user, so it should not be including the header and footer